File: luhn/luhn.go
package luhn

import "unicode"

func Reverse(s string) string {
    n := len(s)
    rs := make([]rune, n)
    for _, r := range s {
        n--
        rs[n] = r
    }
    return string(rs[n:])
}

func Valid(id string) bool {
    double := true
    prep := []int{}
    for _, r := range Reverse(id) {
        if unicode.IsDigit(r) {
            double = !double
            d := int(r-'0')
            if double {
                d *= 2
                if d > 9 {
                    d -= 9
                }
            }
            prep = append(prep, d)
        } else if unicode.IsSpace(r) {
            continue
        } else {
            return false
        }
    }
    if len(prep) < 2 {
        return false
    }
    s := 0
    for _, v := range prep  {
        s += v
    }
    return s % 10 == 0
}

File: python/luhn/luhn.py
class Luhn:
    def __init__(self, card_num: str) -> None:
        self.data = card_num

    def valid(self) -> bool:
        double = True
        prep = []
        for r in self.data[::-1]:
            if r.isdigit():
                double = not double
                d = int(r)
                if double:
                    d *= 2
                    if d > 9:
                        d -= 9
                prep.append(d)
            elif not r.isspace():
                return False
        if len(prep) < 2:
            return False
        return sum(v for v in prep) % 10 == 0
